File: openmesh-apple/MeshFluxIos/core/OpenmeshAppLibBridge.swift
//
//  OpenmeshAppLibBridge.swift
//  MeshFluxIos
//
//  封装 OpenMeshGo 的 OMOpenmeshAppLib，实现 OpenmeshAppLibProtocol，
//  将 Go 返回的类型转换为 App 内使用的 WalletSecretsV1 / VpnStatus。
//

import Foundation
import OpenMeshGo

final class OpenmeshAppLibBridge: OpenmeshAppLibProtocol {
    private let omLib: OMOpenmeshAppLib

    init(omLib: OMOpenmeshAppLib) {
        self.omLib = omLib
    }

    func initApp(_ config: Data) throws {
        try omLib.initApp(config)
    }

    func generateMnemonic12() throws -> String {
        var err: NSError?
        let s = omLib.generateMnemonic12(&err)
        if let e = err { throw e }
        return s
    }

    func createEvmWallet(_ mnemonic: String, password: String) throws -> String {
        var err: NSError?
        let json = omLib.createEvmWallet(mnemonic, password: password, error: &err)
        if let e = err { throw e }
        return json
    }

    func decryptEvmWallet(_ keystoreJSON: String, password: String) throws -> WalletSecretsV1 {
        let om = try omLib.decryptEvmWallet(keystoreJSON, password: password)
        return WalletSecretsV1(
            v: Int(om.v),
            privateKeyHex: om.privateKeyHex,
            address: om.address
        )
    }

    func getTokenBalance(_ address: String, tokenName: String, networkName: String) throws -> String {
        let threadDesc = Thread.isMainThread ? "main" : "background"
        NSLog(
            "OpenmeshAppLibBridge.getTokenBalance start thread=%@ address=%@ token=%@ network=%@",
            threadDesc,
            address,
            tokenName,
            networkName
        )
        var err: NSError?
        let balance = omLib.getTokenBalance(address, tokenName: tokenName, networkName: networkName, error: &err)
        if let e = err {
            NSLog("OpenmeshAppLibBridge.getTokenBalance failed thread=%@ error=%@", threadDesc, String(describing: e))
            throw e
        }
        NSLog("OpenmeshAppLibBridge.getTokenBalance success thread=%@ balance=%@", threadDesc, balance)
        return balance
    }

    func getSupportedNetworks() throws -> String {
        var err: NSError?
        let json = omLib.getSupportedNetworks(&err)
        if let e = err { throw e }
        return json
    }

    func getVpnStatus() -> VpnStatus? {
        guard let om = omLib.getVpnStatus() else { return nil }
        NSLog("OpenmeshAppLibBridge.getVpnStatus connected=%@ server=%@", om.connected.description, om.server)
        return VpnStatus(
            connected: om.connected,
            server: om.server,
            bytesIn: om.bytesIn,
            bytesOut: om.bytesOut
        )
    }
}

File: openmesh-android/app/src/main/java/com/meshnetprotocol/android/core/OpenmeshAppLibBridge.kt
package com.meshnetprotocol.android.core

import android.util.Log

/**
 * OpenMesh Go 库的 Kotlin 桥接实现
 * 封装 gomobile 生成的 Java 类，实现 OpenmeshAppLibProtocol 接口
 * 对应 iOS 的 OpenmeshAppLibBridge
 * 
 * 注意：实际使用时需要等待 AAR 正确生成后，根据 gomobile 生成的真实类名调整此文件
 */
class OpenmeshAppLibBridge(private val omLib: Any) : OpenmeshAppLibProtocol {

    override fun initApp(config: ByteArray) {
        // TODO: 调用实际的 Go 方法
        // omLib.initApp(config)
        Log.d(TAG, "initApp called with config size: ${config.size}")
    }

    override fun generateMnemonic12(): String {
        // TODO: 调用实际的 Go 方法
        // return omLib.generateMnemonic12()
        return "witch collapse practice feed shame open despair creek road again ice least"
    }

    override fun createEvmWallet(mnemonic: String, password: String): String {
        // TODO: 调用实际的 Go 方法
        // return omLib.createEvmWallet(mnemonic, password)
        return "{\"address\":\"0x1234567890abcdef1234567890abcdef12345678\",\"v\":1}"
    }

    override fun decryptEvmWallet(keystoreJSON: String, password: String): WalletSecretsV1 {
        // TODO: 调用实际的 Go 方法
        // val omWallet = omLib.decryptEvmWallet(keystoreJSON, password)
        return WalletSecretsV1(
            v = 1,
            privateKeyHex = "deadbeef1234567890abcdef",
            address = "0x1234567890abcdef1234567890abcdef12345678"
        )
    }

    override fun getTokenBalance(address: String, tokenName: String, networkName: String): String {
        val isMain = Thread.currentThread() == android.os.Looper.getMainLooper().thread
        val threadDesc = if (isMain) "main" else "background"
        Log.d(TAG, "getTokenBalance start thread=$threadDesc address=$address token=$tokenName network=$networkName")
        
        try {
            // TODO: 调用实际的 Go 方法
            // val balance = omLib.getTokenBalance(address, tokenName, networkName)
            val balance = "100.00"
            Log.d(TAG, "getTokenBalance success thread=$threadDesc balance=$balance")
            return balance
        } catch (e: Exception) {
            Log.e(TAG, "getTokenBalance failed thread=$threadDesc error=${e.message}")
            throw e
        }
    }

    override fun getSupportedNetworks(): String {
        // TODO: 调用实际的 Go 方法
        // return omLib.getSupportedNetworks()
        return "[\"base-mainnet\",\"base-testnet\"]"
    }

    override fun getVpnStatus(): VpnStatus? {
        // TODO: 调用实际的 Go 方法
        // val omStatus = omLib.vpnStatus ?: return null
        return VpnStatus(
            connected = false,
            server = "",
            bytesIn = 0,
            bytesOut = 0
        )
    }

    companion object {
        private const val TAG = "OpenmeshAppLibBridge"
    }
}
